src/auth_service/auth_adapter/core/verify_totp.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Application logic for verifying user TOTP tokens."""
+
+from fastapi import HTTPException, status
+from ghga_service_commons.utils.utc_dates import now_as_utc
+from hexkit.protocols.dao import ResourceNotFoundError
+
+from auth_service.user_management.user_registry.models.dto import (
+    StatusChange,
+    UserStatus,
+)
+from auth_service.user_management.user_registry.ports.dao import UserDao
+
+from ..ports.dao import UserToken, UserTokenDao
+from ..ports.session_store import SessionStorePort
+from ..ports.totp import TOTPHandlerPort
+from .session_store import Session, SessionState
+
+__all__ = ["verify_totp"]
+
+
+async def verify_totp(  # noqa: C901, PLR0912, PLR0913
```

Comment:
For this one is quite hard to fully track all states mentally.
As there's a TODO it should be okay for now, but this could hopefully be improved by factoring out the nested middle part, as that seems to deal with a slightly different concern with a different side effect.

Reply:
Yes, it's a bit complex, I'm also not totally happy. We should refactor it later.